src/dawn/Optimizer/StencilFunctionInstantiation.h

```diff
@@ -427,6 +425,8 @@ class StencilFunctionInstantiation {
 
   /// @brief finalizes the binding of the arguments of a stencil function.
   /// In particular it associates new accessIDs of arguments that are nested stencil function calls
+  void closeFunctionBindings(const std::vector<int>&);
```

Comment:
is it necessary to do this on the IIR-Level? why not let the pass handle this?

Reply:
the call to close the function bindings is only ever used by this one pass. Having the functionality in the StencilInstantiation and exposing it to everyone is great for re-use but adds to compile-time and is worse in separation of concerns if it is only used once by one specific instance.

Would it make sense that the pass itself has a function call (in an anonymous namespace) that does exactly this that gets a shared-pointer to a StencilInstantiation as an argument?

Reply:
ok, I see now what you meant, thanks. 
This function is called from within  `finalizeStencilFunctionSetup`, that is also called from StatementMapper. I added all these function to separate different functionality into separate functions, mainly to improve readability. The thing is that StatementMapper is doing lot of stuff, among others instantiating the stencil functions. It is doing low level stuff with the expressions, and so on and very hard to follow. Now, in the temp -> stencil function we have new sir::StencilFunction that we need to instantiate in the same manner. So I had to extract the functionality in methods that we can reuse, and tried to make it readable. For example, since the stencil function is instantiated in different phases, I created a StencilFunctionCandidate that is being parsed and filled and only in the finalize method, the function arguments are bound, and the candidate is promoted from candidate to a stencil function....
In summary, I think we reuse this code, but furthermore improves the readability of it, but maybe have a look again, and let me know otherwise if we can do better